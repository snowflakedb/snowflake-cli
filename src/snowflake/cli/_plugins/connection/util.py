# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import json
import logging
import os
from enum import Enum
from functools import lru_cache
from typing import Any, Dict, Optional

from click.exceptions import ClickException
from snowflake.connector import SnowflakeConnection
from snowflake.connector.cursor import DictCursor

log = logging.getLogger(__name__)


ALLOWLIST_QUERY = "SELECT SYSTEM$ALLOWLIST()"
SNOWFLAKE_DEPLOYMENT = "SNOWFLAKE_DEPLOYMENT"
LOCAL_DEPLOYMENT_REGION: str = "us-west-2"


class MissingConnectionAccountError(ClickException):
    def __init__(self, conn: SnowflakeConnection):
        super().__init__(
            "Could not determine account by system call, configured account name, or configured host. Connection: "
            + repr(conn)
        )


class MissingConnectionRegionError(ClickException):
    def __init__(self, host: str | None):
        super().__init__(
            f"The connection host ({host}) was missing or not in "
            "the expected format "
            "(<account>.<deployment>.snowflakecomputing.com)"
        )


class UIParameter(Enum):
    NA_ENABLE_REGIONLESS_REDIRECT = "UI_SNOWSIGHT_ENABLE_REGIONLESS_REDIRECT"
    NA_EVENT_SHARING_V2 = "ENABLE_EVENT_SHARING_V2_IN_THE_SAME_ACCOUNT"
    NA_ENFORCE_MANDATORY_FILTERS = (
        "ENFORCE_MANDATORY_FILTERS_FOR_SAME_ACCOUNT_INSTALLATION"
    )
    NA_FEATURE_RELEASE_CHANNELS = "FEATURE_RELEASE_CHANNELS"


def get_ui_parameter(
    conn: SnowflakeConnection, parameter: UIParameter, default: Any
) -> Any:
    """
    Returns the value of a single UI parameter.
    If the parameter is not found, the default value is returned.
    """

    ui_parameters = get_ui_parameters(conn)
    return ui_parameters.get(parameter, default)


@lru_cache()
def get_ui_parameters(conn: SnowflakeConnection) -> Dict[UIParameter, Any]:
    """
    Returns the UI parameters from the SYSTEM$BOOTSTRAP_DATA_REQUEST function
    """

    parameters_to_fetch = [param.value for param in UIParameter]

    # Parsing of the Json and the filtering is happening here in Snowflake CLI
    # in order to avoid requiring a warehouse in Snowflake
    query = "call system$bootstrap_data_request('CLIENT_PARAMS_INFO')"
    *_, cursor = conn.execute_string(query)

    json_map = json.loads(cursor.fetchone()[0])

    return {
        UIParameter(row["name"]): row["value"]
        for row in json_map["clientParamsInfo"]
        if row["name"] in parameters_to_fetch
    }


def clear_ui_parameters_cache() -> None:
    """
    Clear the LRU cache for UI parameters to ensure fresh state.

    This function should be called between tests or operations to prevent
    state pollution from cached UI parameters. It clears the cache for
    get_ui_parameters function.
    """
    get_ui_parameters.cache_clear()


def is_regionless_redirect(conn: SnowflakeConnection) -> bool:
    """
    Determines if the deployment this connection refers to uses
    regionless URLs in Snowsight (/orgname/account) or regional URLs
    (/region/account). If we cannot determine the correct value we
    assume it's regionless, as this is true for most production deployments.
    """
    try:
        return get_ui_parameter(conn, UIParameter.NA_ENABLE_REGIONLESS_REDIRECT, True)
    except:
        log.warning(
            "Cannot determine regionless redirect; assuming True.", exc_info=True
        )
        return True


def get_host_region(host: str) -> str | None:
    """
    Looks for hosts of form
    <account>.[x.y.z].snowflakecomputing.com
    Returns the three-part [region identifier] or None.
    """
    host_parts = host.split(".")
    if host_parts[-1] == "local":
        return LOCAL_DEPLOYMENT_REGION
    elif len(host_parts) == 6:
        return ".".join(host_parts[1:4])
    return None


def guess_regioned_host_from_allowlist(conn: SnowflakeConnection) -> str | None:
    """
    Use SYSTEM$ALLOWLIST to find a regioned host (<account>.x.y.z.snowflakecomputing.com)
    that corresponds to the given Snowflake connection object.
    """
    try:
        *_, cursor = conn.execute_string(ALLOWLIST_QUERY, cursor_class=DictCursor)
        allowlist_tuples = json.loads(cursor.fetchone()["SYSTEM$ALLOWLIST()"])
        for t in allowlist_tuples:
            if t["type"] == SNOWFLAKE_DEPLOYMENT:
                if get_host_region(t["host"]) is not None:
                    return t["host"]
    except:
        log.warning(
            "Could not call SYSTEM$ALLOWLIST; returning an empty guess.", exc_info=True
        )
    return None


def get_region(conn: SnowflakeConnection) -> str:
    """
    Get the region of the given connection, or raise MissingConnectionRegionError.
    """
    if conn.host:
        if region := get_host_region(conn.host):
            return region

    if host := guess_regioned_host_from_allowlist(conn):
        if region := get_host_region(host):
            return region

    raise MissingConnectionRegionError(host or conn.host)


def get_context(conn: SnowflakeConnection) -> str:
    """
    Determines the first part of the path in a Snowsight URL.
    This could be a region or it could be an organization, depending
    on whether or not the underlying deployment uses regionless URLs.
    """
    if is_regionless_redirect(conn):
        *_, cursor = conn.execute_string(
            f"select system$return_current_org_name()", cursor_class=DictCursor
        )
        return cursor.fetchone()["SYSTEM$RETURN_CURRENT_ORG_NAME()"]

    return get_region(conn)


def get_account(conn: SnowflakeConnection) -> str:
    """
    Determines the account that this connection refers to.
    """
    try:
        *_, cursor = conn.execute_string(
            f"select current_account_name()", cursor_class=DictCursor
        )
        return cursor.fetchone()["CURRENT_ACCOUNT_NAME()"].lower()
    except Exception as e:
        # try to extract the account from the connection information
        if conn.account:
            return conn.account

        if conn.host:
            host_parts = conn.host.split(".")
            return host_parts[0]

        raise MissingConnectionAccountError(conn)


def get_snowsight_host(conn: SnowflakeConnection) -> str:
    try:
        *_, cursor = conn.execute_string(
            f"select system$get_snowsight_host()", cursor_class=DictCursor
        )
        return cursor.fetchone()["SYSTEM$GET_SNOWSIGHT_HOST()"]
    except Exception as e:
        # if we cannot determine the host, assume we're on prod
        return "https://app.snowflake.com"


def make_snowsight_url(conn: SnowflakeConnection, path: str) -> str:
    """
    Returns a URL on the correct Snowsight instance for the connected account.
    The path that is passed in must already be properly URL-encoded, and
    can optionally contain a hash/fragment (e.g. #).

    See also identifier_for_url.
    """
    snowsight_host = get_snowsight_host(conn)
    deployment = get_context(conn)
    account = get_account(conn)
    path_with_slash = path if path.startswith("/") else f"/{path}"
    return f"{snowsight_host}/{deployment}/{account}{path_with_slash}"


def strip_if_value_present(value: Optional[str]) -> Optional[str]:
    return value.strip() if value else value


def ensure_that_path_exist(path: Optional[str]) -> Optional[str]:
    if path and not os.path.exists(path):
        raise ClickException(f"Path {path} does not exist.")
    return path


def strip_and_check_if_exists(value: Optional[str]) -> Optional[str]:
    return ensure_that_path_exist(strip_if_value_present(value))
