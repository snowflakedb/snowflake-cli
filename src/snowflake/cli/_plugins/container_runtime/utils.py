import enum
from dataclasses import dataclass
from pathlib import PurePath
from typing import Any, Dict, List, Optional, Union, cast

from packaging import version
from snowflake.snowpark import session


class SnowflakeCloudType(enum.Enum):
    AWS = "aws"
    AZURE = "azure"
    GCP = "gcp"

    @classmethod
    def from_value(cls, value: str) -> "SnowflakeCloudType":
        assert value
        for k in cls:
            if k.value == value.lower():
                return k
        else:
            raise ValueError(f"'{cls.__name__}' enum not found for '{value}'")


@dataclass(frozen=True)
class PayloadEntrypoint:
    file_path: PurePath
    main_func: Optional[str]


@dataclass(frozen=True)
class UploadedPayload:
    # TODO: Include manifest of payload files for validation
    stage_path: PurePath
    entrypoint: List[Union[str, PurePath]]


@dataclass(frozen=True)
class ComputeResources:
    cpu: float  # Number of vCPU cores
    memory: float  # Memory in GiB
    gpu: int = 0  # Number of GPUs
    gpu_type: Optional[str] = None


@dataclass(frozen=True)
class ImageSpec:
    repo: str
    image_name: str
    image_tag: str
    resource_requests: ComputeResources
    resource_limits: ComputeResources

    @property
    def full_name(self) -> str:
        return f"{self.repo}/{self.image_name}:{self.image_tag}"


def get_current_region_id(sess: session.Session) -> str:
    res = session.sql(sess, "SELECT CURRENT_REGION() AS CURRENT_REGION").collect()[0]

    return cast(str, res.CURRENT_REGION)


def get_current_snowflake_version(
    sess: session.Session, *, statement_params: Optional[Dict[str, Any]] = None
) -> version.Version:
    """Get Snowflake Version as a version.Version object follow PEP way of versioning, that is to say:
        "7.44.2 b202312132139364eb71238" to <Version('7.44.2+b202312132139364eb71238')>

    Args:
        sess: Snowpark Session.
        statement_params: Statement params. Defaults to None.

    Returns:
        The version of Snowflake Version.
    """
    res = session.sql(
        sess,
        "SELECT CURRENT_VERSION() AS CURRENT_VERSION",
        statement_params=statement_params,
    ).collect()[0]

    version_str = res.CURRENT_VERSION
    assert isinstance(version_str, str)

    version_str = "+".join(version_str.split())
    return version.parse(version_str)
