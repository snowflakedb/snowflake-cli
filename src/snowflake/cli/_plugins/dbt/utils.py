# Copyright (c) 2025 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from snowflake.cli._plugins.dbt.constants import KNOWN_SUBCOMMANDS


def _extract_dbt_args(args: list[str]) -> list[str]:
    flags = set()

    for arg in args:
        if arg.startswith("-"):
            if "=" in arg:
                flag_name = arg.split("=", 1)[0]
                flags.add(flag_name)
            else:
                flags.add(arg)
        elif arg in KNOWN_SUBCOMMANDS:
            flags.add(arg)

    return sorted(list(flags))
