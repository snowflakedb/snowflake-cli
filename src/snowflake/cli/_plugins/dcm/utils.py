# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


def format_test_failures(
    failed_expectations: list, total_tests: int, failed_count: int
) -> str:
    """Format test failures into a nice error message."""
    lines = [
        "Failed expectations:",
    ]

    for failed in failed_expectations:
        table_name = failed.get("table_name", "Unknown")
        expectation_name = failed.get("expectation_name", "Unknown")
        metric_name = failed.get("metric_name", "Unknown")
        expectation_expr = failed.get("expectation_expression", "N/A")
        value = failed.get("value", "N/A")

        lines.append(f"  Table: {table_name}")
        lines.append(f"  Expectation: {expectation_name}")
        lines.append(f"  Metric: {metric_name}")
        lines.append(f"  Expression: {expectation_expr}")
        lines.append(f"  Actual value: {value}")
        lines.append("")

    passed_tests = total_tests - failed_count
    lines.append(
        f"Tests completed: {passed_tests} passed, {failed_count} failed out of {total_tests} total."
    )

    return "\n".join(lines)


def format_refresh_results(refreshed_tables: list) -> str:
    """Format refresh results into a concise user-friendly message."""
    if not refreshed_tables:
        return "No dynamic tables found in the project."

    total_tables = len(refreshed_tables)
    refreshed_count = sum(
        1 for table in refreshed_tables if table.get("refreshed_dt_count", 0) > 0
    )
    up_to_date_count = total_tables - refreshed_count

    return f"{refreshed_count} dynamic table(s) refreshed. {up_to_date_count} dynamic table(s) up-to-date."
