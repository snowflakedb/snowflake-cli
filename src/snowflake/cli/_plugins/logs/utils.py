from datetime import datetime
from typing import List, NamedTuple, Optional, Tuple

from snowflake.cli.api.exceptions import CliArgumentError, CliSqlError
from snowflake.connector.cursor import SnowflakeCursor

LOG_LEVELS = ["TRACE", "DEBUG", "INFO", "WARN", "ERROR", "FATAL"]

LogsQueryRow = NamedTuple(
    "LogsQueryRow",
    [
        ("timestamp", datetime),
        ("database_name", str),
        ("schema_name", str),
        ("object_name", str),
        ("log_level", str),
        ("log_message", str),
    ],
)


def sanitize_logs(logs: SnowflakeCursor | List[Tuple]) -> List[LogsQueryRow]:
    try:
        return [LogsQueryRow(*log) for log in logs]
    except TypeError:
        raise CliSqlError(
            "Logs table has incorrect format. Please check the logs_table in your database"
        )


def get_timestamp_query(from_time: Optional[datetime], to_time: Optional[datetime]):
    if from_time and to_time and from_time > to_time:
        raise CliArgumentError(
            "From_time cannot be later than to_time. Please check the values"
        )
    query = []

    if from_time is not None:
        query.append(f"AND timestamp >= TO_TIMESTAMP_LTZ('{from_time.isoformat()}')\n")

    if to_time is not None:
        query.append(f"AND timestamp <= TO_TIMESTAMP_LTZ('{to_time.isoformat()}')\n")

    return "".join(query)


def get_log_levels(log_level: str):
    if log_level.upper() not in LOG_LEVELS and log_level != "":
        raise CliArgumentError(
            f"Invalid log level. Please choose from {', '.join(LOG_LEVELS)}"
        )

    if log_level == "":
        log_level = "INFO"

    return LOG_LEVELS[LOG_LEVELS.index(log_level.upper()) :]


def parse_log_levels_for_query(log_level: str):
    return ", ".join(f"'{level}'" for level in get_log_levels(log_level))
