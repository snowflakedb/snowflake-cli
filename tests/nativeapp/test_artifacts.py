# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from pathlib import Path

import pytest
from click import ClickException
from snowflake.cli._plugins.nativeapp.artifacts import (
    VersionInfo,
    build_bundle,
    find_events_definitions_in_manifest_file,
    find_version_info_in_manifest_file,
)
from snowflake.cli.api.artifacts.common import (
    ArtifactError,
    DeployRootError,
    NotInDeployRootError,
    SourceNotFoundError,
    TooManyFilesError,
)
from snowflake.cli.api.project.definition import load_project
from snowflake.cli.api.project.schemas.entities.common import PathMapping
from snowflake.cli.api.project.util import to_identifier
from yaml import safe_dump

from tests.nativeapp.factories import ManifestFactory
from tests.nativeapp.utils import (
    assert_dir_snapshot,
)
from tests.testing_utils.files_and_dirs import pushd, temp_local_dir


@pytest.mark.parametrize("project_definition_files", ["napp_project_1"], indirect=True)
def test_napp_project_1_artifacts(project_definition_files, os_agnostic_snapshot):
    project_root = project_definition_files[0].parent
    native_app = load_project(project_definition_files).project_definition.native_app

    with pushd(project_root) as local_path:
        deploy_root = Path(local_path, native_app.deploy_root)
        build_bundle(local_path, deploy_root, native_app.artifacts)

        assert_dir_snapshot(deploy_root.relative_to(local_path), os_agnostic_snapshot)

        # we should be able to re-bundle without any errors happening
        build_bundle(local_path, deploy_root, native_app.artifacts)

        # any additional files created in the deploy root will be obliterated by re-bundle
        with open(deploy_root / "unknown_file.txt", "w") as handle:
            handle.write("I am an unknown file!")
        assert_dir_snapshot(deploy_root.relative_to(local_path), os_agnostic_snapshot)

        build_bundle(local_path, deploy_root, native_app.artifacts)
        assert_dir_snapshot(deploy_root.relative_to(local_path), os_agnostic_snapshot)


@pytest.mark.parametrize("project_definition_files", ["napp_project_1"], indirect=True)
def test_source_not_found(project_definition_files):
    project_root = project_definition_files[0].parent
    with pytest.raises(SourceNotFoundError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[PathMapping(src="NOTFOUND.md", dest="NOTFOUND.md")],
        )


@pytest.mark.parametrize("project_definition_files", ["napp_project_1"], indirect=True)
def test_glob_matched_nothing(project_definition_files):
    project_root = project_definition_files[0].parent
    with pytest.raises(SourceNotFoundError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[PathMapping(src="**/*.jar", dest=".")],
        )


@pytest.mark.parametrize("project_definition_files", ["napp_project_1"], indirect=True)
def test_outside_deploy_root_three_ways(project_definition_files):
    project_root = project_definition_files[0].parent
    with pytest.raises(NotInDeployRootError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[PathMapping(src="setup.sql", dest="..")],
        )

    with pytest.raises(NotInDeployRootError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[PathMapping(src="setup.sql", dest="/")],
        )

    with pytest.raises(NotInDeployRootError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[PathMapping(src="app", dest=".")],
        )


@pytest.mark.parametrize("project_definition_files", ["napp_project_1"], indirect=True)
def test_bad_deploy_root(project_definition_files):
    project_root = project_definition_files[0].parent
    with pytest.raises(DeployRootError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "..", "deploy"),
            artifacts=[],
        )

    with pytest.raises(DeployRootError):
        with open(project_root / "deploy", "w") as handle:
            handle.write("Deploy root should not be a file...")

        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[],
        )


@pytest.mark.parametrize("project_definition_files", ["napp_project_1"], indirect=True)
def test_no_artifacts(project_definition_files):
    project_root = project_definition_files[0].parent
    with pytest.raises(ArtifactError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[],
        )


@pytest.mark.parametrize("project_definition_files", ["napp_project_1"], indirect=True)
def test_too_many_files(project_definition_files):
    project_root = project_definition_files[0].parent
    with pytest.raises(TooManyFilesError):
        build_bundle(
            project_root,
            deploy_root=Path(project_root, "deploy"),
            artifacts=[
                PathMapping(
                    src="app/streamlit/*.py", dest="somehow_combined_streamlits.py"
                )
            ],
        )


@pytest.mark.parametrize(
    "label", [None, "", "label 'nested' quotes", "with$pecial?", "with space"]
)
@pytest.mark.parametrize("patch_name", [None, "1", 42])
@pytest.mark.parametrize(
    "version_name", [None, "v1", "1", 2, "1.2", 1.3, "1.2.3", "0.x", "foo", "abc def"]
)
def test_find_version_info_in_manifest_file(version_name, patch_name, label):
    manifest_contents = {"manifest_version": 1, "version": {}}

    if all(value is None for value in (version_name, patch_name, label)):
        manifest_contents.pop("version")
    if version_name is not None:
        manifest_contents["version"]["name"] = version_name
    if patch_name is not None:
        manifest_contents["version"]["patch"] = patch_name
    if label is not None:
        manifest_contents["version"]["label"] = label

    deploy_root_structure = {"manifest.yml": safe_dump(manifest_contents)}
    with temp_local_dir(deploy_root_structure) as deploy_root:
        v, p, l = find_version_info_in_manifest_file(deploy_root=deploy_root)
        if version_name is None:
            assert v is None
        else:
            assert v == to_identifier(str(version_name))

        if patch_name is None:
            assert p is None
        else:
            assert p == int(patch_name)

        if label is None:
            assert l is None
        else:
            assert l == label


@pytest.mark.parametrize("version_info", ["some_name", ["li1", "li2"], "", 4])
def test_bad_version_info_in_manifest_file_throws_error(version_info):
    manifest_contents = ManifestFactory(version=version_info)

    deploy_root_structure = {"manifest.yml": manifest_contents}
    with temp_local_dir(deploy_root_structure) as deploy_root:
        with pytest.raises(ClickException) as err:
            find_version_info_in_manifest_file(deploy_root=deploy_root)
        assert (
            err.value.message
            == "Error occurred while reading manifest.yml. Received unexpected version format."
        )


def test_read_empty_version_info_in_manifest_file():
    manifest_contents = ManifestFactory(
        version={"name": None, "patch": None, "label": None}
    )

    deploy_root_structure = {"manifest.yml": manifest_contents}
    with temp_local_dir(deploy_root_structure) as deploy_root:

        version_info = find_version_info_in_manifest_file(deploy_root=deploy_root)
        assert version_info == VersionInfo(
            version_name=None, patch_number=None, label=None
        )


def test_read_empty_version_in_manifest_file():
    manifest_contents = ManifestFactory(version=None)

    deploy_root_structure = {"manifest.yml": manifest_contents}
    with temp_local_dir(deploy_root_structure) as deploy_root:

        version_info = find_version_info_in_manifest_file(deploy_root=deploy_root)
        assert version_info == VersionInfo(
            version_name=None, patch_number=None, label=None
        )


@pytest.mark.parametrize(
    "configuration_section, expected_output",
    [
        [
            None,
            [],
        ],
        [
            "test",
            [],
        ],
        [
            ["test"],
            [],
        ],
        [
            {},
            [],
        ],
        [
            {"telemetry_event_definitions": "test"},
            [],
        ],
        [
            {"telemetry_event_definitions": {"key": "value"}},
            [],
        ],
        [
            {"telemetry_event_definitions": []},
            [],
        ],
        [
            {
                "telemetry_event_definitions": [
                    {"type": "ERRORS_AND_WARNINGS", "sharing": "MANDATORY"},
                    {"type": "DEBUG_LOGS", "sharing": "OPTIONAL"},
                ]
            },
            [
                {
                    "name": "SNOWFLAKE$ERRORS_AND_WARNINGS",
                    "type": "ERRORS_AND_WARNINGS",
                    "sharing": "MANDATORY",
                },
                {
                    "name": "SNOWFLAKE$DEBUG_LOGS",
                    "type": "DEBUG_LOGS",
                    "sharing": "OPTIONAL",
                },
            ],
        ],
    ],
)
def test_find_events_in_manifest_file(configuration_section, expected_output):
    manifest_contents = {"manifest_version": 1, "version": {"name": "v1", "patch": 1}}
    manifest_contents["configuration"] = configuration_section

    deploy_root_structure = {"manifest.yml": safe_dump(manifest_contents)}
    with temp_local_dir(deploy_root_structure) as deploy_root:
        assert (
            find_events_definitions_in_manifest_file(deploy_root=deploy_root)
            == expected_output
        )
