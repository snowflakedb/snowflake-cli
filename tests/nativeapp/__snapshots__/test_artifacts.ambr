# serializer version: 1
# name: test_napp_project_1_artifacts[napp_project_1]
  '''
  d output/deploy
  d output/deploy/app
  f output/deploy/app/README.md
  f output/deploy/setup.sql
  d output/deploy/ui
  f output/deploy/ui/config.py
  f output/deploy/ui/main.py
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].1
  '''
  ===== Contents of: output/deploy/app/README.md =====
  app/README.md
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].10
  '''
  ===== Contents of: output/deploy/unknown_file.txt =====
  I am an unknown file!
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].11
  '''
  d output/deploy
  d output/deploy/app
  f output/deploy/app/README.md
  f output/deploy/setup.sql
  d output/deploy/ui
  f output/deploy/ui/config.py
  f output/deploy/ui/main.py
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].12
  '''
  ===== Contents of: output/deploy/app/README.md =====
  app/README.md
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].13
  '''
  ===== Contents of: output/deploy/setup.sql =====
  create versioned schema myschema;
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].14
  '''
  ===== Contents of: output/deploy/ui/config.py =====
  # config.py
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].15
  '''
  ===== Contents of: output/deploy/ui/main.py =====
  # main.py
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].2
  '''
  ===== Contents of: output/deploy/setup.sql =====
  create versioned schema myschema;
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].3
  '''
  ===== Contents of: output/deploy/ui/config.py =====
  # config.py
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].4
  '''
  ===== Contents of: output/deploy/ui/main.py =====
  # main.py
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].5
  '''
  d output/deploy
  d output/deploy/app
  f output/deploy/app/README.md
  f output/deploy/setup.sql
  d output/deploy/ui
  f output/deploy/ui/config.py
  f output/deploy/ui/main.py
  f output/deploy/unknown_file.txt
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].6
  '''
  ===== Contents of: output/deploy/app/README.md =====
  app/README.md
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].7
  '''
  ===== Contents of: output/deploy/setup.sql =====
  create versioned schema myschema;
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].8
  '''
  ===== Contents of: output/deploy/ui/config.py =====
  # config.py
  
  '''
# ---
# name: test_napp_project_1_artifacts[napp_project_1].9
  '''
  ===== Contents of: output/deploy/ui/main.py =====
  # main.py
  
  '''
# ---
